web/base_viewer.js

```diff
@@ -396,6 +396,51 @@ class BaseViewer {
     throw new Error("Not implemented: _setDocumentViewerElement");
   }
 
+  /**
+   * @private
+   */
+  get _setDocumentAllowFetchPages() {
```

Comment:
Having a getter that is called `set` is a bit confusing. How about a regular method that is something like `autoFetchDelay()`?

Reply:
> Having a getter that is called set is a bit confusing.

That was simply to sort-of "tie" the functionality to the `setDocument` method, similar to the exisiting `_setDocumentViewerElement` getter.

> How about a regular method that is something like `autoFetchDelay()`?

Out of curiosity: Is there a reason that a regular method would be preferred over a getter, e.g. related to performance?
Also, how about simply `allowFetchPages` for the name instead? At least to me a name such as "autoFetchDelay" suggests that something will always load, however that's not actually guaranteed here (e.g. in the `disableAutoFetch=true` case).

Reply:
I also find `_setDocumentViewerElement` pretty confusing. 

There's not really a perf difference or hard fast rule, it's more I prefer that getters be relatively simple property accessors vs a method signals that there's going to be more "work" going on when I call it.

Good point on the autoFetch term. `_allowFetchPages` implies i could do something like `if (_allowFetchPages())`, which also feels a bit off too. 

Before we brainstorm some more names, what about if we changed the viewer to not only wait for an request animation frame, but race it with a requestIdleCallback with the timeout argument?

Reply:
e.g. in https://github.com/mozilla/pdf.js/blob/c3f4690bde8137d80c74203b1ad91476fc2ca160/src/display/api.js#L2870

Reply:
> [...] it's more I prefer that getters be relatively simple property accessors vs a method signals that there's going to be more "work" going on when I call it.

Well, I can't really argue with that :-)

> what about if we changed the viewer to not only wait for an request animation frame, but race it with a requestIdleCallback with the timeout argument?

Sorry, but I'm not entirely sure that I follow your train of thought here; mind elaborating?

Also, cross-browser support for `requestIdleCallback` doesn't seem to be that great according to https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback#Browser_compatibility (since we still have to support both Safari and IE11), so you'd have to rely on polyfills.

> https://github.com/mozilla/pdf.js/blob/c3f4690bde8137d80c74203b1ad91476fc2ca160/src/display/api.js#L2870

Again, I'm sure that I understand how that relates to the viewer...

Reply:
Part of the reason we don't ever resolve the onePageRendered promise is because we wait for a request animation frame, which we never get when the pdf viewer is hidden (we'd still also have to handle no pages being visible issue). After thinking about this some more though, we'd be rendering a page that isn't visible and doing work we don't really need to do.

Looking at telemetry, a lot of PDFs can take more than a 1 second to render, so I'd pref we handle the above bugs with without racing promises and a timeout.

Can we instead do something like:
```
onePageRenderedOrForceFetch() {
  if (!this.container.offsetParent || 
        !this._location || /* not sure why this is here below */
        this._getVisiblePages().views.length === 0) {
    return Promise.resolve();
  }
  return this._onePageRenderedCapability.promise;
}
```

Reply:
> [...] a lot of PDFs can take more than a 1 second to render, so I'd pref we handle the above bugs with without racing promises and a timeout.

The "viewer hidden"-case is easy to handle (with just `this.container.offsetParent`), but the "viewer visible, but all pages hidden"-case is more complicated.
I simply wanted to avoid doing more work *every single time* that the viewer loads normally, with pages being visible, since [`getVisibleElements`](https://github.com/mozilla/pdf.js/blob/master/web/ui_utils.js#L441) may be a bit expensive in long document (thousands of pages).

> Can we instead do something like:
> 
> ```
> onePageRenderedOrForceFetch() {
>   if (!this.container.offsetParent || 
>         !this._location || /* not sure why this is here below */
>         this._getVisiblePages().views.length === 0) {
>     return Promise.resolve();
>   }
>   return this._onePageRenderedCapability.promise;
> }
> ```

If we do this we're *forced* to always call `this._getVisiblePages()` every time that the viewer loads, just to fix https://bugzilla.mozilla.org/show_bug.cgi?id=1618955 which should hopefully be a *very* rare use-case.

Reply:
Could we do a cheaper check to see if just the page we want to render is visible?

Reply:
> Could we do a cheaper check to see if just the page we want to render is visible?

Sure we can, it just took me a little while to figure out a good/safe way of doing so!
I've implemented this as a special case in `getVisibleElements`, such that we'll check the first page immediately but still fallback to the normal binary search to hopefully prevent any future errors.

The latest version of the patch thus implements your suggestion from https://github.com/mozilla/pdf.js/pull/11646#discussion_r396621823; thank you!

Reply:
Sorry, I managed to made a *really* stupid mistake when reading the existing code!

Apparently I completely missed this part of the `binarySearchFirstItem` helper function: https://github.com/mozilla/pdf.js/blob/85838fc50507f4a98d97930b4ba8b40f388d39a4/web/ui_utils.js#L228-L237
Obviously that code already contains a fast-path for the *first* element, which means that just calling `this._getVisiblePages()` immediately after the pages have been initialized should be completely fine (since the first page is always visible initially).

Again, sorry about the churn here!
The latest version of the patch implement your suggestion from https://github.com/mozilla/pdf.js/pull/11646#discussion_r396621823; thank you!